rfcs/20210510-binary-size-reduction/ISA_extension.md

```diff
@@ -0,0 +1,117 @@
+...
+```
+
+Based on this `BUILD_XXX` macros, the following internal macros would be set:
+
+```cpp
+// src/common/impl_registration.hpp:
+
+#if BUILD_PRIMITIVE_ISA_ALL
+#define REG_ISA_AMX(...) __VA_ARGS__
+#define REG_ISA_AVX512(...) __VA_ARGS__
+#define REG_ISA_AVX2(...) __VA_ARGS__
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#elif BUILD_AVX512
+#define REG_ISA_AMX(...)
+#define REG_ISA_AVX512(...) __VA_ARGS__
+#define REG_ISA_AVX2(...) __VA_ARGS__
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#elif BUILD_AVX2
+#define REG_ISA_AMX(...)
+#define REG_ISA_AVX512(...)
+#define REG_ISA_AVX2(...) __VA_ARGS__
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#elif BUILD_SSE41
+#define REG_ISA_AMX(...)
+#define REG_ISA_AVX512(...)
+#define REG_ISA_AVX2(...)
+#define REG_ISA_SSE41(...) __VA_ARGS__
+#endif
+```
+
+Based on this set of macros, `CPU_INSTANCE_X64` would be extended with the
+following list:
+```cpp
+// src/cpu/cpu_engine.hpp:
+
+#define CPU_INSTANCE_X64_AVX2(...) REG_ISA_AVX2(DNNL_X64_ONLY(CPU_INSTANCE(__VA_ARGS__)))
+#define CPU_INSTANCE_X64_AVX512(...) REG_ISA_AVX512(DNNL_X64_ONLY(CPU_INSTANCE(__VA_ARGS__)))
+#define CPU_INSTANCE_X64_AMX(...) REG_ISA_AMX(DNNL_X64_ONLY(CPU_INSTANCE(__VA_ARGS__)))
+```
+
+Finally, these internal macros are applied inside primitive implementation
+lists:
+
+```cpp
+// src/cpu/cpu_convolution_list.cpp:
+const std::map<conv_impl_key_t, std::vector<pd_create_f>> impl_list_map {
+    {{forward, s8, s8, f32}, {
+        REG_CONV_P_FWD(REG_IP_P_FWD(CPU_INSTANCE_X64(ip_convolution_fwd_t)))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AMX(jit_avx512_core_amx_1x1_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX512(jit_avx512_core_amx_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX512(jit_avx512_core_x8s8s32x_1x1_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX512(jit_avx512_core_x8s8s32x_convolution_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX2(jit_uni_x8s8s32x_1x1_convolution_fwd_t<avx2>))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64_AVX2(jit_uni_x8s8s32x_convolution_fwd_t<avx2>))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64(jit_uni_x8s8s32x_1x1_convolution_fwd_t<sse41>))
+        REG_CONV_P_FWD(CPU_INSTANCE_X64(jit_uni_x8s8s32x_convolution_fwd_t<sse41>))
+        CPU_INSTANCE_AARCH64(jit_sve_512_x8s8s32x_convolution_fwd_t<s8, f32>)
+        REG_CONV_P_FWD(CPU_INSTANCE(_gemm_x8s8s32x_convolution_fwd_t<s8, f32>))
+        REG_CONV_P_FWD(CPU_INSTANCE(ref_convolution_int8_fwd_t))
+        REG_CONV_P_FWD(CPU_INSTANCE(ref_fused_convolution_fwd_t))
+        nullptr,
+    }},
+```
+
+Additionally, to overcome some of [2], instantiations of same classes for
+different ISA should be wrapped with macros as well. Otherwise, those instances
+would still present in the final binary:
+
+```cpp
+// src/cpu/x64/jit_uni_x8s8s32x_convolution.cpp:
+
+...
+REG_ISA_SSE41(template struct jit_uni_x8s8s32x_convolution_fwd_t<sse41>);
+REG_ISA_AVX2(template struct jit_uni_x8s8s32x_convolution_fwd_t<avx2>);
```

Comment:
I'm not sure if it's super useful, just want share an idea on how to hide some logic. We can introduce 1 common helper macro to redirect instantiations to the proposed isa specific macros. The downside is that isa name in the macros name should be in lower case.
  
```cpp
// impl - implementation type
// isa - isa name without suffixes
// isa suffix and any other remaining things that are needed to instantiate the impl
#define REG_INSTANCE(impl, isa, ...) REG_ISA_##isa(template struct impl<isa##__VA_ARGS__>)

// Simple example when there is only isa name
REG_INSTANCE(jit_uni_x8s8s32x_convolution_fwd_t, avx2);
// More complex example when there is isa name and its suffix
REG_INSTANCE(brgemm_convolution_fwd_t, avx512, _core_bf16_amx_bf16);
// Even more complex case, when there isa name, its suffix and data types
REG_INSTANCE(jit_uni_dw_convolution_fwd_t, avx512, _core, bf16, bf16);
```
The output will be the following:
```cpp
    REG_ISA_avx2(template struct jit_uni_x8s8s32x_convolution_fwd_t<avx2>);
    REG_ISA_avx512(template struct brgemm_convolution_fwd_t<avx512_core_bf16_amx_bf16>);
    REG_ISA_avx512(template struct jit_uni_dw_convolution_fwd_t<avx512_core, bf16, bf16>);
```

Reply:
Looks scary if I'm not the one who needs to push some update in the list... :)

Reply:
I thought using a single macro `REG_INSTANCE` could be less error prone because in this case you cannot do something like this by mistake: `REG_ISA_AVX512(template struct jit_uni_x8s8s32x_convolution_fwd_t<avx2>)`. 

As for the scary part, I think the explanation of the `REG_INSTANCE` macro above is pretty much descriptive so I wouldn't expect people will struggle with using it. Maybe I'm missing some other downsides of the approach. Please share if you see some.

Reply:
Propose to move discussion about implementation details for actual PR once done. I will consider your comments when crafting a solution. Thanks.